src/irmin-pack/unix/chunked_suffix.ml

```diff
@@ -232,15 +254,42 @@ module Make (Io : Io.S) (Errs : Io_errors.S with module Io = Io) = struct
       Ao.open_ro ~path ~end_poff ~dead_header_size
     in
     let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
-    { inventory; root; dead_header_size }
+    { start_idx; inventory; root; dead_header_size }
 
+  let start_idx t = t.start_idx
+  let chunk_num t = Inventory.count t.inventory
   let appendable_ao t = (Inventory.appendable t.inventory).ao
   let end_poff t = appendable_ao t |> Ao.end_poff
   let length t = Inventory.length t.inventory
 
   let read_exn t ~off ~len buf =
-    let chunk, poff = Inventory.find ~off t.inventory in
-    Ao.read_exn chunk.ao ~off:poff ~len buf
+    let rec read progress_off suffix_off len_requested =
+      let open Int63.Syntax in
+      (* Find chunk with [suffix_off] and calculate length we can read. *)
+      let chunk, poff = Inventory.find ~off:suffix_off t.inventory in
+      let chunk_end_poff = Ao.end_poff chunk.ao in
+      let read_end_poff = poff + len_requested in
+      let len_read =
+        if read_end_poff > chunk_end_poff then chunk_end_poff - poff
+        else len_requested
+      in
+
+      (* Perform read. If this is the first read, we can use [buf]; otherwise,
+         we create a new buffer and transfer after the read. *)
+      let len_i = Int63.to_int len_read in
+      let is_first_read = progress_off = Int63.zero in
+      let ao_buf = if is_first_read then buf else Bytes.create len_i in
```

Comment:
wdyt about not distinguishing between the first and the subsequent reads? and always create a new buffer which is transferred after the read into [buf]?

Reply:
I had that at first. The reason I added this (pre?) optimization is that the vast majority of reads will be within one chunk (only GC currently reads across chunks) and so I wanted to avoid the overhead of unnecessary buffer allocation.